packaging/public-api/build.gradle.kts

```diff
@@ -30,15 +31,76 @@ dependencies {
     externalApi(libs.groovy)
     // Required to inject services into tasks and other objects
     externalApi(libs.inject)
-    // Moslty used for nullability annotations
-    externalApi(libs.jsr305)
+    // JSpecify is part of the public API for nullability annotations
     externalApi(libs.jspecify)
-    // We use this to annotate type parameters as @Nullable
+    // We don't use this anymore for annotating the public API, but we support plugin types annotated with it
+    externalApi(libs.jsr305)
+    // We don't use this anymore in public types, but it is still part of the public API
     externalApi(libs.jetbrainsAnnotations)
     // SLF4J logging is part of our public API
     externalApi(libs.slf4jApi)
     // We only need this because of AntTarget :o
     externalApi(libs.ant)
+
+    // Modules that are part of the legacy gradleApi() dependency
+    // See DependencyClassPathProvider and DependencyClassPathNotationConverter
+    legacyExternalApi(libs.groovyAnt)
+    legacyExternalApi(libs.groovyAstbuilder)
+    legacyExternalApi(libs.groovyDatetime)
+    legacyExternalApi(libs.groovyDateUtil)
+    legacyExternalApi(libs.groovyDoc)
+    legacyExternalApi(libs.groovyJson)
+    legacyExternalApi(libs.groovyNio)
+    legacyExternalApi(libs.groovyTemplates)
+    legacyExternalApi(libs.groovyXml)
+    legacyExternalApi(libs.kotlinStdlib)
+    legacyExternalApi(libs.kotlinReflect)
+    legacyExternalApi(libs.nativePlatform)
+    legacyExternalApi(libs.log4jToSlf4j)
+}
+// These are relocated impldeps that we should exclude
+configurations.legacyExternalApi {
+    extendsFrom(configurations.externalApi.get())
+    exclude(module = "qdox")
+    exclude(module = "javaparser-core")
+}
+
+// The JAR built by this module includes the ABI of internals
+val internalBaseName = gradleModule.identity.baseName.map { "$it-internal" }
+tasks.jarGradleApi {
+    archiveBaseName = internalBaseName
+}
+tasks.classpathManifest {
+    manifestFile = layout.buildDirectory.dir("generated-resources/classpath-manifest")
+        .zip(internalBaseName) { dir, baseName ->
+            dir.file("$baseName-classpath.properties")
+        }
+}
+configurations {
+    gradleApiElements {
+        outgoing {
+            capability(internalBaseName.map { "$group:$it:$version" })
+        }
+    }
+}
+
+// Legacy Gradle API
+val legacyBaseName = gradleModule.identity.baseName.map { "$it-legacy" }
+tasks.jarGradleApiLegacy {
+    archiveBaseName = legacyBaseName
+}
+tasks.legacyClasspathManifest {
+    manifestFile = layout.buildDirectory.dir("generated-resources/legacy-classpath-manifest")
+        .zip(legacyBaseName) { dir, baseName ->
+            dir.file("$baseName-classpath.properties")
+        }
+}
+configurations {
+    legacyGradleApiElements {
+        outgoing {
+            capability(legacyBaseName.map { "$group:$it:$version" })
+        }
```

Comment:
How come we're configuring all of this in the buildscript? It seems like most if not all of this logic could be moved to the plugin and configured where the tasks/configurations are created